x-pack/plugin/rollup/src/main/java/org/elasticsearch/xpack/rollup/action/TransportPutRollupJobAction.java

```diff
@@ -147,16 +149,18 @@ private static RollupJob createRollupJob(RollupJobConfig config, ThreadPool thre
         return new RollupJob(config, filteredHeaders);
     }
 
+    @SuppressWarnings("unchecked")
     static void createIndex(RollupJob job, ActionListener<AcknowledgedResponse> listener,
                             PersistentTasksService persistentTasksService, Client client, Logger logger) {
 
-        String jobMetadata = "\"" + job.getConfig().getId() + "\":" + job.getConfig().toJSONString();
+        Map<String, Object> jobMetaData = toMap(job.getConfig().toJSONString());
+        Map<String, Object> mapping = toMap(Rollup.DYNAMIC_MAPPING_TEMPLATE);
 
-        String mapping = Rollup.DYNAMIC_MAPPING_TEMPLATE
-                .replace(Rollup.MAPPING_METADATA_PLACEHOLDER, jobMetadata);
+        ((Map<String, Object>)((Map<String, Object>) ((Map<String, Object>) ((Map<String, Object>) mapping.get("mappings"))
```

Comment:
This is...uhh... not ideal, but seemed less hacky than doing a string replace.  I could be convinced otherwise, open to alternatives.

Reply:
I wonder if we should use an `XContentBuilder` directly to simplify the understanding of this code. Something like:
````
private static XContentBuilder createMappings(RollupJobConfig config) throws IOException {
        return XContentBuilder.builder(XContentType.JSON.xContent())
            .startObject("mappings")
                .startObject("_doc")
                    .startObject("_meta")
                        .field("rollup-version", Version.CURRENT)
                        .startObject("_rollup")
                            .field(config.getId(), config)
                        .endObject()
                    .endObject()
                    .startArray("dynamic_templates")
                        .startObject("strings")
                            .field("match_mapping_type", "string")
                            .startObject("mapping")
                                .field("type", "keyword")
                            .endObject()
                        .endObject()
                        .startObject("date_histograms")
                            .field("path_match", "*.date_histogram.timestamp")
                            .startObject("mapping")
                                .field("type", "date")
                            .endObject()
                        .endObject()
                    .endArray()
                .endObject()
            .endObject();
    }
````
... seems easier to read imo. We could also retrieve the `dynamic_templates` from a file like we do today but we have only 2 rules so I am not too concerned by the inlining here. What do you think ?

Reply:
This looks a hundred times better :)  Will make the change